renderdoc/driver/metal/metal_init_state.cpp

```diff
@@ -22,14 +22,77 @@
  * THE SOFTWARE.
  ******************************************************************************/
 
+#include "metal_buffer.h"
 #include "metal_common.h"
 #include "metal_device.h"
 
+static rdcliteral NameOfType(MetalResourceType type)
+{
+  switch(type)
+  {
+    case eResBuffer: return "MTLBuffer"_lit;
+    default: break;
+  }
+  return "MTLResource"_lit;
+}
+
 bool WrappedMTLDevice::Prepare_InitialState(WrappedMTLObject *res)
 {
   ResourceId id = GetResourceManager()->GetID(res);
 
   MetalResourceType type = res->m_Record->m_Type;
+
+  if(type == eResBuffer)
+  {
+    WrappedMTLBuffer *buffer = (WrappedMTLBuffer *)res;
+    MTL::Buffer *mtlBuffer = Unwrap(buffer);
+    MTL::Buffer *mtlSharedBuffer = NULL;
+    MTL::StorageMode storageMode = mtlBuffer->storageMode();
+    size_t len = mtlBuffer->length();
+    byte *data = NULL;
+    if(storageMode == MTL::StorageModeShared)
+    {
+      // MTLStorageModeShared buffers are automatically synchronized
+      data = (byte *)mtlBuffer->contents();
```

Comment:
For context (you don't have to change anything), on other drivers the initial contents snapshot is generally left as a buffer on the GPU side (the metal equivalent looks like it would be a StorageModePrivate buffer). This defers the readback until such a point that we know we need the data, and avoids the need for synchronisation where possible. It's not a requirement necessarily and it is possible to readback during prepare here - for integrated GPUs which are common on mac I believe it's possibly a better approach aside from the need for repeated synchronisation.

Reply:
Captured here : https://github.com/Zorro666/renderdoc/issues/88